File: python/844-Backspace-String-Compare.py
import itertools


# region Two Pointers Pythonic
# Time O(m + n)
# Space O(1)
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def myFunc(u: str):
            skips = 0
            for char in reversed(u):
                if char == "#":
                    skips += 1
                elif skips:
                    skips -= 1
                else:
                    yield char

        return all(x == y for x, y in itertools.zip_longest(myFunc(s), myFunc(t)))


# endregion


# region Two Pointers
# Time O(m + n) where n and m are the lengths of the strings
# Space O(1)
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        i, j = len(s) - 1, len(t) - 1
        skipS, skipT = 0, 0

        while i >= 0 or j >= 0:
            while i >= 0:
                if s[i] == "#":
                    skipS += 1
                    i -= 1
                elif skipS > 0:
                    skipS -= 1
                    i -= 1
                else:
                    break

            while j >= 0:
                if t[j] == "#":
                    skipT += 1
                    j -= 1
                elif skipT > 0:
                    skipT -= 1
                    j -= 1
                else:
                    break

            if i >= 0 and j >= 0 and s[i] != t[j]:
                return False

            if (i >= 0) != (j >= 0):
                return False
            i -= 1
            j -= 1
        return True


# endregion


# region String Builder Solution
# Time O(n + m) where n and m are the lengths of the strings
# Space O(n + m)
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def build(input: str) -> str:
            chars = []
            for elem in input:
                if elem != "#":
                    chars.append(elem)
                elif chars:
                    chars.pop()
            return "".join(chars)

        return build(s) == build(t)


# endregion

File: c#/844-Backspace-String-Compare.cs
// Medium Solution
// Two pointers, walk through string backwards.
// O(1) Space, O(M + N) Time where M & N are respective lengths
public class Solution
{
    public bool BackspaceCompare(string s, string t)
    {
        int sPtr = s.Length - 1,
            tPtr = t.Length - 1;
        int sSkips = 0,
            tSkips = 0;

        while (sPtr >= 0 || tPtr >= 0)
        {
            while (sPtr >= 0)
            {
                if (s[sPtr] == '#')
                {
                    sSkips++;
                    sPtr--;
                }
                else if (sSkips > 0)
                {
                    sSkips--;
                    sPtr--;
                }
                else
                    break;
            }

            while (tPtr >= 0)
            {
                if (t[tPtr] == '#')
                {
                    tSkips++;
                    tPtr--;
                }
                else if (tSkips > 0)
                {
                    tSkips--;
                    tPtr--;
                }
                else
                    break;
            }

            // If we're in the middle of both strings, chars should be equal
            if (sPtr >= 0 && tPtr >= 0 && s[sPtr] != t[tPtr])
                return false;

            // Trying to compare char to null should end loop
            if ((sPtr >= 0) != (tPtr >= 0))
                return false;

            sPtr--;
            tPtr--;
        }

        return true;
    }
}

// Easy Solution
// Walk forward through strings and build new ones with backspaces applied, then compare
// O(M + N) Space, O(M + N) Time, where M & N are respective lengths

public class Solution
{
    private string applyBackspaces(string s)
    {
        var sb = new StringBuilder();

        foreach (var c in s)
        {
            if (c == '#' && sb.Length > 0)
                sb.Remove(sb.Length - 1, 1);
            else if (c != '#')
                sb.Append(c);
            Console.WriteLine(sb);
        }
        return sb.ToString();
    }

    public bool BackspaceCompare(string s, string t)
    {
        s = applyBackspaces(s);
        t = applyBackspaces(t);

        return string.Equals(s, t);
    }
}
